The latest version does not support go1.18 compilation!!

### Does your log mention database corruption?

If your Syncthing log reports panics because of database corruption it is
most likely a fault with your system's storage or memory. Affected log
entries will contain lines starting with `panic: leveldb`. You will need to
delete the index database to clear this, by running `syncthing
-reset-database`.

### Include required information

Please be sure to include at least:

 - which version of Syncthing and what operating system you are using
 - browser and version, if applicable
 - what happened,
 - what you expected to happen instead, and
 - any steps to reproduce the problem.

If it helps, the context of this report is in relation to OpenWrt: https://github.com/openwrt/packages/issues/18267

The master branch recently updated the Go lang host version to 1.18, several Go lang packages have been broken by this, most with the following error:

> ../../vendor/github.com/lucas-clemente/quic-go/internal/qtls/go118.go:5:13: cannot use "quic-go doesn't build on Go 1.18 yet." (untyped string constant) as int value in variable declaration

You can view the compile failure via the faillogs: https://downloads.openwrt.org/snapshots/faillogs/x86_64/packages/syncthing/compile.txt

The current package version in OpenWrt 1.18.2, I tried building 1.19.2 but go other errors:

> Building targets
github.com/marten-seemann/qtls-go1-18
# github.com/marten-seemann/qtls-go1-18
../../vendor/github.com/marten-seemann/qtls-go1-18/cipher_suites.go:143:43: undeclared name: any (requires version go1.18 or later)
../../vendor/github.com/marten-seemann/qtls-go1-18/cipher_suites.go:413:45: undeclared name: any (requires version go1.18 or later)
../../vendor/github.com/marten-seemann/qtls-go1-18/cipher_suites.go:418:46: undeclared name: any (requires version go1.18 or later)
../../vendor/github.com/marten-seemann/qtls-go1-18/cipher_suites.go:426:45: undeclared name: any (requires version go1.18 or later)
../../vendor/github.com/marten-seemann/qtls-go1-18/conn.go:177:10: undeclared name: any (requires version go1.18 or later)
../../vendor/github.com/marten-seemann/qtls-go1-18/conn.go:183:13: undeclared name: any (requires version go1.18 or later)
../../vendor/github.com/marten-seemann/qtls-go1-18/conn.go:211:62: undeclared name: any (requires version go1.18 or later)
../../vendor/github.com/marten-seemann/qtls-go1-18/conn.go:973:14: undeclared name: any (requires version go1.18 or later)
../../vendor/github.com/marten-seemann/qtls-go1-18/conn.go:1056:33: undeclared name: any (requires version go1.18 or later)
../../vendor/github.com/marten-seemann/qtls-go1-18/common.go:1496:41: undeclared name: any (requires version go1.18 or later)
../../vendor/github.com/marten-seemann/qtls-go1-18/conn.go:1056:33: too many errors

So we have a build issue with Go 1.18 currently.

Hope that clarifies the original report.

--- a/build.go
+++ b/build.go
@@ -883,7 +883,7 @@ func shouldRebuildAssets(target, srcdir string) bool {
 
 func updateDependencies() {
 	runPrint(goCmd, "get", "-u", "./cmd/...")
-	runPrint(goCmd, "mod", "tidy", "-go=1.16", "-compat=1.16")
+	runPrint(goCmd, "mod", "tidy", "-go=1.17", "-compat=1.17")
 
 	// We might have updated the protobuf package and should regenerate to match.
 	proto()

--- a/go.mod
+++ b/go.mod
@@ -57,4 +57,27 @@ require (
 	google.golang.org/protobuf v1.27.1
 )
 
-go 1.16
+require (
+	github.com/Azure/go-ntlmssp v0.0.0-20200615164410-66371956d46c // indirect
+	github.com/beorn7/perks v1.0.1 // indirect
+	github.com/cheekybits/genny v1.0.0 // indirect
+	github.com/go-ole/go-ole v1.2.6 // indirect
+	github.com/go-task/slim-sprig v0.0.0-20210107165309-348f09dbbbc0 // indirect
+	github.com/golang/protobuf v1.5.2 // indirect
+	github.com/marten-seemann/qtls-go1-16 v0.1.4 // indirect
+	github.com/marten-seemann/qtls-go1-17 v0.1.0 // indirect
+	github.com/marten-seemann/qtls-go1-18 v0.1.0-beta.1 // indirect
+	github.com/matttproud/golang_protobuf_extensions v1.0.1 // indirect
+	github.com/nxadm/tail v1.4.8 // indirect
+	github.com/onsi/ginkgo v1.16.4 // indirect
+	github.com/oschwald/maxminddb-golang v1.8.0 // indirect
+	github.com/petermattis/goid v0.0.0-20180202154549-b0b1615b78e5 // indirect
+	github.com/power-devops/perfstat v0.0.0-20210106213030-5aafc221ea8c // indirect
+	github.com/prometheus/client_model v0.2.0 // indirect
+	github.com/russross/blackfriday/v2 v2.1.0 // indirect
+	github.com/yusufpapurcu/wmi v1.2.2 // indirect
+	golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1 // indirect
+	gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7 // indirect
+)
+
+go 1.17

